compiler/compiler.go

```diff
@@ -1,210 +1,136 @@
-// result. It is assumed that the parse result is semantically
-// correct and would compile with protoc.
+// Package compiler creates FileDescriptorSets from a *.proto input and
+// FileDescriptors for *parser.Proto input. It is assumed that the
+// parse result is semantically correct and would compile with protoc.
 package compiler
 
 import (
 	"fmt"
-	"strings"
+	"io"
+	"os"
+	"path/filepath"
 
 	"github.com/alecthomas/protobuf/parser"
 	pb "google.golang.org/protobuf/types/descriptorpb"
 )
 
-type fdBuilder struct {
-	proto3   bool
-	fileDesc *pb.FileDescriptorProto
-	types    types
-	pkg      string
-}
-
-type messageBuilder struct {
-	proto3      bool
-	messageDesc *pb.DescriptorProto
-	types       types
-	scope       []string
-}
-
-func NewFileDescriptor(file string, p *parser.Proto) *pb.FileDescriptorProto {
-	b := &fdBuilder{
-		proto3: p.Syntax == "proto3",
-		pkg:    protoPkg(p),
-	}
-	b.types = analyseTypes(p)
-	b.buildFileDescriptor(file, p)
-	return b.fileDesc
-}
-
-func (b *fdBuilder) buildFileDescriptor(file string, p *parser.Proto) {
-	var syntax *string
-	if b.proto3 {
-		s := "proto3"
-		syntax = &s
-	}
-	b.fileDesc = &pb.FileDescriptorProto{
-		Name:   &file,
-		Syntax: syntax,
-	}
-	for _, e := range p.Entries {
-		b.addEntry(e)
-	}
-}
-
-func (b *fdBuilder) addEntry(e *parser.Entry) {
-	switch {
-	case e.Package != "":
-		b.fileDesc.Package = &e.Package
-	case e.Import != nil:
-		b.buildImports(e.Import)
-	case e.Message != nil:
-		scope := []string{}
-		if b.pkg != "" {
-			scope = append(scope, b.pkg)
+func NewFileDescriptorSet(files, importPaths []string, includeImports bool) (*pb.FileDescriptorSet, error) {
+	done := map[string]bool{}
+	origFiles := map[string]bool{}
+	for _, file := range files {
+		origFiles[file] = true
+	}
+	protos, err := readProtos(files, importPaths, done)
+	if err != nil {
+		return nil, err
+	}
+	types := newTypes(protos)
+	fds := &pb.FileDescriptorSet{}
+	for _, p := range protos {
+		if includeImports || origFiles[p.file] {
+			fd := NewFileDescriptor(p, types)
+			fds.File = append(fds.File, fd)
 		}
-		m := newMessage(e.Message, b.proto3, scope, b.types)
-		b.fileDesc.MessageType = append(b.fileDesc.MessageType, m)
-	case e.Enum != nil:
-	case e.Service != nil:
-	case e.Option != nil:
-	case e.Extend != nil:
-	default:
-		panic(fmt.Sprintf("%s: cannot interpret Entry", e.Pos))
 	}
+	return fds, nil
 }
 
-func (b *fdBuilder) buildImports(imp *parser.Import) {
-	b.fileDesc.Dependency = append(b.fileDesc.Dependency, imp.Name)
-	if imp.Public {
-		idx := int32(len(b.fileDesc.Dependency) - 1)
-		b.fileDesc.PublicDependency = append(b.fileDesc.PublicDependency, idx)
-	}
-}
-
-func newMessage(m *parser.Message, proto3 bool, scope []string, types types) *pb.DescriptorProto {
-	b := &messageBuilder{
-		proto3: proto3,
-		scope:  append(scope, m.Name),
-		types:  types,
-	}
-	b.buildMessageDescriptor(m)
-	return b.messageDesc
+type Proto struct {
```

Comment:
Never used  outside the package? Make it private?

Reply:
yeah - naming is hard. it was private and clashed with `import "google.golang.org/protobuf/proto"`.
i think i am a little rusty in this domain - what would be a good name?
I'll rename it to `ast` - please let me know if that's completely off.